acme/challenge.go

```diff
@@ -463,6 +501,333 @@ func deviceAttest01Validate(ctx context.Context, ch *Challenge, db DB, jwk *jose
+		}
+	}
+
+	// extract and validate pubArea, sig, certInfo and alg properties from the request body
+	pubArea, ok := att.AttStatement["pubArea"].([]byte)
+	if !ok {
+		return nil, NewError(ErrorBadAttestationStatementType, "invalid pubArea in attestation statement")
+	}
+	if len(pubArea) == 0 {
+		return nil, NewError(ErrorBadAttestationStatementType, "pubArea is empty")
+	}
+
+	sig, ok := att.AttStatement["sig"].([]byte)
+	if !ok {
+		return nil, NewError(ErrorBadAttestationStatementType, "invalid sig in attestation statement")
+	}
+	if len(sig) == 0 {
+		return nil, NewError(ErrorBadAttestationStatementType, "sig is empty")
+	}
+
+	certInfo, ok := att.AttStatement["certInfo"].([]byte)
+	if !ok {
+		return nil, NewError(ErrorBadAttestationStatementType, "invalid certInfo in attestation statement")
+	}
+	if len(certInfo) == 0 {
+		return nil, NewError(ErrorBadAttestationStatementType, "certInfo is empty")
+	}
+
+	alg, ok := att.AttStatement["alg"].(int64)
+	if !ok {
+		return nil, NewError(ErrorBadAttestationStatementType, "invalid alg in attestation statement")
+	}
+
+	var hash crypto.Hash
+	switch alg {
+	case -257: // RS256
+		hash = crypto.SHA256
+	case -8: // EdDSA
+		hash = crypto.Hash(0)
+	case -7: // ES256
+		hash = crypto.SHA256
+	default:
+		return nil, NewError(ErrorBadAttestationStatementType, "invalid alg %d in attestation statement", alg)
+	}
+
+	// recreate the generated key certification parameter values and verify
+	// the attested key using the public key of the AK.
+	certificationParameters := &attest.CertificationParameters{
+		Public:            pubArea,  // the public key that was attested
+		CreateAttestation: certInfo, // the attested properties of the key
+		CreateSignature:   sig,      // signature over the attested properties
+	}
+	verifyOpts := attest.VerifyOpts{
+		Public: akCert.PublicKey, // public key of the AK that attested the key
+		Hash:   hash,
+	}
+	if err = certificationParameters.Verify(verifyOpts); err != nil {
+		return nil, WrapError(ErrorBadAttestationStatementType, err, "invalid certification parameters")
+	}
+
+	// decode the "certInfo" data. This won't fail, as it's also done as part of Verify().
+	tpmCertInfo, err := tpm2.DecodeAttestationData(certInfo)
+	if err != nil {
+		return nil, WrapError(ErrorBadAttestationStatementType, err, "failed decoding attestation data")
+	}
+
+	keyAuth, err := KeyAuthorization(ch.Token, jwk)
+	if err != nil {
+		return nil, WrapError(ErrorBadAttestationStatementType, err, "failed creating key auth digest")
+	}
+	hashedKeyAuth := sha256.Sum256([]byte(keyAuth))
+
+	// verify the WebAuthn object contains the expect key authorization digest, which is carried
+	// within the encoded `certInfo` property of the attestation statement.
+	if subtle.ConstantTimeCompare(hashedKeyAuth[:], []byte(tpmCertInfo.ExtraData)) == 0 {
+		return nil, NewError(ErrorBadAttestationStatementType, "key authorization does not match")
+	}
```

Comment:
Is this how this is verified in a TPM? With a YubiKey we verify with the public key in the certificate that `sig` matches keyAuth.

Also, will the last comparison work if EdDSA is used? No hardware, afaik supports it, but you're handling this case above.

Reply:
Removed the `EdDSA` case; you're right that it probably doesn't make sense. It will now return with an unsupported algorithm. 

When an application key is attested by an AK while including `QualifyingData`, that data is included with the certification data. The certified creation data is available in the `CreateAttestation` blob that's part of the `CertificationParameters`. When the input `CertificationParameters` have been verified (signed by AK, proper parameters, expected properties, etc) and the `CreateAttestation` data blob is decoded, the `QualifyingData` is present in the `ExtraData` property. The properties of the key that are checked include the data was created by a TPM, the key was generated by a TPM, cannot be exported, and some more.

The TPM spec mentions this about the property:
```
extraData | TPM2B_DATA | external info supplied by caller (often in qualifyingData parameter)

NOTE A TPM2B_DATA structure provides room for a digest and a method
indicator to indicate the components of the digest. The definition of
this method indicator is outside the scope of this specification.
```